src/mongo/db/storage/csv_file.cpp

```diff
@@ -0,0 +1,338 @@
+#include <fstream>  // IWYU pragma: keep
+#include <string>
+#include <sys/stat.h>
+
+#include "mongo/base/error_codes.h"
+#include "mongo/base/string_data.h"
+#include "mongo/db/query/query_knobs_gen.h"
+#include "mongo/db/storage/csv_file.h"
+#include "mongo/db/storage/io_error_message.h"
+#include "mongo/util/assert_util.h"
+#include "mongo/util/time_support.h"
+
+namespace mongo {
+using namespace fmt::literals;
+
+CsvFileInput::CsvFileInput(const std::string& fileRelativePath, const std::string& metaFileRelative)
+    : _fileAbsolutePath((externalPipeDir == "" ? kDefaultPipePath : externalPipeDir) +
+                        fileRelativePath),
+      _metadataAbsolutePath((externalPipeDir == "" ? kDefaultPipePath : externalPipeDir) +
+                            metaFileRelative),
+      _ifs() {
+    uassert(200000400,
+            "File path must not include '..' but {} does"_format(_fileAbsolutePath),
+            _fileAbsolutePath.find("..") == std::string::npos);
+}
+
+CsvFileInput::~CsvFileInput() {
+    _ifs.close();
+}
+
+void CsvFileInput::doOpen() {
+    std::string metadata;
+    _ifs.open(_metadataAbsolutePath.c_str(), std::ios::binary | std::ios::in);
+    uassert(ErrorCodes::FileNotOpen,
+            "File has failed to open. Make sure the file exists: "_format(
+                getErrorMessage("open", _metadataAbsolutePath)),
+            _ifs.is_open());
+    std::getline(_ifs, metadata);
+
+    _md = getMetadata(parseLine(metadata));
+    _ifs.close();
+
+    _ifs.open(_fileAbsolutePath.c_str(), std::ios::binary | std::ios::in);
+
+    uassert(ErrorCodes::FileNotOpen,
+            "error = {}"_format(getErrorMessage("open", _fileAbsolutePath)),
+            _ifs.is_open());
+}
+
+// caller must ensure that buffer size is greater than or equal to the size of the bsonObject
+// to be returned. If not, it would throw exception (Not enough size in buffer)
+int CsvFileInput::doRead(char* data, int size) {
+
+    auto bsonObj = readBsonObj();
+
+    if (!bsonObj) {
+        return 0;
+    }
+
+    int nRead = bsonObj->objsize();
+    uassert(200000402, "Buff Size too small", nRead <= size);
+    memcpy(data, bsonObj->objdata(), nRead);
+    return nRead;
+}
+
+void CsvFileInput::doClose() {
+    _ifs.close();
+}
+
+bool CsvFileInput::isOpen() const {
+    return _ifs.is_open();
+}
+
+bool CsvFileInput::isGood() const {
+    return _ifs.good();
+}
+
+bool CsvFileInput::isFailed() const {
+    return _ifs.fail();
+}
+
+bool CsvFileInput::isEof() const {
+    return _ifs.eof();
+}
+
+// assuming that header is returned by parseLine, which means, each of its element is
+// name of the field with type name
+Metadata CsvFileInput::getMetadata(const std::vector<std::string>& header) {
+    Metadata ret;
+
+    for (auto& field : header) {
+        int index = field.find('/');
+        int len = field.length() - index - 1;
+
+        auto fieldName = field.substr(0, index);
+        auto typeName = field.substr(index + 1, len);
```

Comment:
Should we check whether `typeName` is empty string, meaning that `len` is 0? Are we testing this scenario?

Reply:
Done